Namespace filter for service discovery (issue #1000).

mbialkowski1 opened the pull request.

wind57 commented on spring-cloud-kubernetes-integration-tests/spring-cloud-kubernetes-fabric8-client-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/configmap/Fabric8DiscoveryNamespaceFilterIT.java:
```diff
@@ -0,0 +1,248 @@
+/**
+ * @author mbialkowski1
+ */
+class Fabric8DiscoveryNamespaceFilterIT {
```
this is fabulous! I'd like to ask for an integration test in k8s-client also, the reason for that is that eventually we will move away from fabric8 for discovery (just like we did for reload for example). again, this is awesome and thank you!

ryanjbaxter commented:
Done!

wind57 commented:
I'l leaving a note here for myself mainly, we need this change in `KubernetesCatalogWatch` too. The fabric8 one + when we add a k8s-native implementation of catalog watcher, too.

mbialkowski1 commented:
> @mbialkowski1 tell us when this is done and I will gladly take a look (most probably Ryan too) thank you

I think I'm done, I added an integration test for `org.springframework.cloud.kubernetes.discovery.KubernetesDiscoveryClient`.
I also found the problem in the method `org.springframework.cloud.kubernetes.client.discovery.KubernetesInformerDiscoveryClient#getInstances` 
We use only one service:
```
V1Service service = properties.allNamespaces() ? this.serviceLister.list().stream()
				.filter(svc -> serviceId.equals(svc.getMetadata().getName())).findFirst().orElse(null)
```

If a service with the same name is in several different namespaces then we lost information about it.
I corrected it, can you see if it is ok? (https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1113/commits/fed3e4f74a581ad1455096a61849e76ba9974c26)
What else should I do/improve in this PR?

wind57 commented:
I will take a look at this in the upcoming weeks, I just want to work (and clarify some things first), if you do not mind.

ryanjbaxter reviewed (commented):
Do we still retain the functionality that is `all-namespace` is false and no specific namespace is specified we still use the current namespace?

mbialkowski1 commented:
> Do we still retain the functionality that is `all-namespace` is false and no specific namespace is specified we still use the current namespace?

Yes, it should work like that.
